_posts/2023/2023-05-16-generative-ai.html

```diff
@@ -0,0 +1,240 @@
+	code produced when developers use Copilot versus when they do not. Their
+	results suggest that using Copilot may produce code with significantly
+	slower performance.
+      </li>
+      <li>
+	The study by Dakhel et al. [<a href="#ref-11">11</a>] suggests that
+	Copilot can provide solutions for most fundamental algorithmic problems
+	with the exception of some cases that yield buggy results. They also
+	find that human programmers generate a higher ratio of correct solutions
+	than Copilot. This contradicts the auxiliary findings of Erhabor
+	[<a href="#ref-7">7</a>].
+      </li>
+      <li>
+	Even OpenAI's own report on GPT-4 [<a href="#ref-12">12</a>] indicates
+	that approximately 74% of all medium-difficulty and 93% of all
+	hard-difficulty problems in Leetcode could not be solved.
+      </li>
+    </ol>
+    <p>
+      The jury is therefore still out, even before we worry about the murky
+      legalities of training models on people's data without their prior
+      knowledge or consent. However:
+    </p>
+    <ol>
+      <li>
+	These models are still Question and Answer type models, not independent
+	agents: they need people to prompt them to generate code responses. The
+	need for proper prompt engineering is starting to be widely recognized
+	[<a href="#ref-13">13</a>, <a href="#ref-14">14</a>]. This suggests that
+	requirements engineering, long neglected, will become increasingly
+	critical.
+      </li>
+      <li>
+	While generative AI can write some code reasonably well, it's hard to
+	know whether the code is right or wrong. We
+	can <a href="https://www.fuzzingbook.org/">fuzz test</a> to see if the
+	application crashes, but we cannot know if the functional requirements
+	are met.  We therefore need humans to see if the code does what is
+	needed, rather than just what was asked for.
+      </li>
+      <li>
+	Current generative AI models cannot plan ahead [<a href="#ref-8">8</a>],
+	and the coding tasks that have been tested with generative AI are quite
+	small. We therefore still need humans to break problems down into
+	smaller pieces so that models can generate code.
+      </li>
+      <li>
+	Finally, as even OpenAI states [<a href="#ref-12">12</a>], models "do
+	not learn from experience". Given this limitation, when a new version of
+	a library is released to fix a bug or to add a feature, generative models
+	won't be able to make use of them without retraining. (Of course, the
+	same can be said of people…)
+      </li>
+    </ol>
+    <p>
+      So, can generative AI automate away software engineering jobs?  The answer
+      right now is, "No, but they are still useful."  And, as my
+      colleague <a href="https://asokan.org/asokan/">Asokan</a> notes, while
+      generative AI is not going to automate away SE jobs, it is certainly going
+      to change the way software engineers work. If someone's expertise is
+      writing code, they will almost certainly be replaced by generative AI in
```

Comment:
If someone's expertise is writing code -> If someone's expertise is only writing code